### PR TITLE
4.08: Fix exceptions leaking from Parser

### PR DESCRIPTION
**Pull request by let-def**
This is a tentative fix for #972.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,6 +6,10 @@ build:
 workspace:
 	dune build --workspace=dune-workspace.template
 
+workspace-ocamlmerlin:
+	dune build --workspace=dune-workspace.template \
+		src/frontend/ocamlmerlin_server.exe
```
Comment (trefis) [inline Makefile]: This feels out of place in this PR. Also, I'd suggest building "merlin.install" instead of "src/frontend/ocamlmerlin_server.exe").

Comment (let-def) [reply Makefile]: Right.

File: src/ocaml/preprocess/dune [hunk 2]
```diff
@@ -1,10 +1,26 @@
...
+  if need_cpp then
+    ("*.mlyp",
+     "(rule\n\
+       (targets parser_raw.mly)\n\
+       (deps    parser_raw.mlyp)\n\
+       (action\n\
+         (with-stdout-to %{targets}\n\
+           (run gcc -E -x c -P %{deps}))))")
+  else 
+    ("*.mly", "")
```
Comment (trefis) [inline src/ocaml/preprocess/dune]: Cute.

**Comment by let-def** [comment]
This patch:
- removes all `error` rules in the grammar (they use the built-in error handling system of yacc which conflicts with merlin recovery algorithm)
- fixes a few conflicts in the grammar (inlining of `match_cases`, precedence annotation for `LET_LWT`)
- replaces all (most) exceptions thrown by a call to `Msupport_parsing.raise_error` to register the error and then returns a sane fallback value.

**Pull request merged**